hilti/include/rt/type-info.h

```diff
@@ -0,0 +1,994 @@
+// Copyright (c) 2020 by the Zeek Project. See LICENSE for details.
+
+#pragma once
+
+#include <optional>
+#include <string>
+#include <variant>
+#include <vector>
+
+#include <hilti/rt/exception.h>
+#include <hilti/rt/fmt.h>
+#include <hilti/rt/types/all.h>
+#include <hilti/rt/types/bytes.h>
+#include <hilti/rt/types/stream.h>
+
+namespace hilti::rt {
+
+struct TypeInfo;
+
+namespace type_info {
+
+/**
+ * Class representing a HILTI value generically through a pair of (1) a raw
+ * pointer refering the value's storage, and (2) type information describing
+ * how to interpret the raw pointer. An instance may be in an invalid state
+ * if there's no underlying value available (e.g., when dereferencing an
+ * unset `optional`).
+ */
+class Value {
```

Comment:
I am not a fan of this type as it just unsafely wraps data owned elsewhere, but could e.g., be stored away obscuring the data dependency. It seems we use it to access data inside `ParsedUnit` and to recursively iterate over values.

I think an implementation where we do not store not-owned data away would be much easier. I do not have a full alternative design ready, but I could imagine something which could still store the type info as e.g., a `shared_ptr<TypeInfo>` (might need to look out for reference cycles, but that could be an unlikely problem in a tree structure), but not store a pointer to the underlying value storage, but instead provide a function to visit a `ParsedUnit`'s data (might need to be a `friend` of `ParsedUnit`). Storing an additional offset into the data wouldn't be safe either and it would be better if we could use safe hilti-rt container access instead.

Such a change would largely be confined to our implementations of types in `detail` here and apart from accessing the underlying storage `Value` could keep most of its interface.

Reply:
As discussed offline, not sure we could build the API much else without taking a very different approach. But I tried the idea of adding back pointers to the top-level value so that we can catch life-time violations. I just pushed that into a new commit a the very end, see what you think about that.

Reply:
Thanks, comments for inline.